### PR TITLE
Release 31

### PR DESCRIPTION
**Pull request by orndorffgrant**
# Release 31

* **Pro Client Representative:** @orndorffgrant 
* **Sponsor Reviewer:** @paride 
* **SRU Reviewer:** @basak 

The goal is to do all of the review on github this time. Please let me know if you're missing anything here.

This is ready to start reviewing but their are some things left to do before it is fully ready.
- [x] Resolve differences between this branch and `main`
- [x] Add changelog entry for v31
- [x] Create SRU bug and add reference to changelog
- [x] fix `fuzzy` translations (fyi @lucasmoura @renanrodrigo)
- [x] #2887 
- [x] #2937 

## Interesting things to note
* The target of this PR, `release`, is actually 12 commits ahead of `ubuntu/devel` in git-ubuntu. You can see them here: https://github.com/canonical/ubuntu-pro-client/commits/release/
* Version 30.1 was only released on noble, so this commit will also be new for xenial, bionic, focal, jammy, mantic: https://github.com/canonical/ubuntu-pro-client/commit/2dec7a182d5af18d804758ec038dd02c067c6b37

### CODE REVIEW

**Comment by renanrodrigo** [comment]
@orndorffgrant I added a commit fixing the fuzzy translations.
@lucasmoura please take a look in this commit.

File: debian/po/pt_BR.po [hunk 1]
```diff
@@ -824,22 +823,22 @@ msgid ""
...
 msgid ""
 "Failed to automatically attach to an Ubuntu Pro subscription {num_attempts} "
 "time(s).\n"
 "The failure was due to: {reason}.\n"
 "The next attempt is scheduled for {next_run_datestring}.\n"
 "You can try manually with `sudo pro auto-attach`."
 msgstr ""
-"{num_attempts} falha(s) ao vincular automaticamente aos serviços do Ubuntu "
-"Pro.\n"
+"{num_attempts} falha(s) ao tentar vincular automaticamente a uma assinatura "
```
Comment (lucasmoura) [inline debian/po/pt_BR.po:834]: Do we really need `tentar` here ?

File: debian/po/pt_BR.po [hunk 2]
```diff
@@ -848,8 +847,8 @@ msgid ""
 "ubuntu-advantage-tools`\n"
 "You can try manually with `sudo pro auto-attach`."
 msgstr ""
-"{num_attempts} falha(s) ao vincular automaticamente aos serviços do Ubuntu "
-"Pro.\n"
+"{num_attempts} falha(s) ao tentar vincular automaticamente a uma assinatura "
```
Comment (lucasmoura) [inline debian/po/pt_BR.po:850]: Same comment here

**Review by lucasmoura — COMMENTED** [review]
@renanrodrigo just a couple of nits for the translation. Let me know if you agree

File: debian/po/pt_BR.po (hunk 1 above)
Comment (renanrodrigo) [reply debian/po/pt_BR.po:834]: I thought it reads better, we tried and failed :D
Can roll back to the old phrase though

Comment (renanrodrigo) [reply debian/po/pt_BR.po]: Done

**Review by lucasmoura — APPROVED** [review]

**Review by paride — COMMENTED** [review]
This basically LGTM; I added a couple of nitty comments on debian/* changes, but speaking of packaging that's it.

There is however one thing I'd like to raise: when I build the package on my machine (Noble with no /etc/apt/sources.list, as I migrated it to a deb833 /etc/apt/sources.list.d/ubuntu.sources), a couple of tests fail. These do not fail when building the package in a Noble schroot. Relevant log:

```
=================================== FAILURES ===================================
_ TestHandleRequiredPackages.test_handle_required_packages[required_packages_directive2-expected_apt_update_calls2-expected_apt_install_calls2-True] _

self = <uaclient.entitlements.tests.test_base.TestHandleRequiredPackages object at 0x7e73a201a150>
m_entitlement_cfg = <PropertyMock name='entitlement_cfg' id='139035066564304'>
m_apt_update = <MagicMock name='run_apt_update_command' id='139035066559888'>
m_apt_install = <MagicMock name='run_apt_install_command' id='139035066564048'>
m_update_sources_list = <MagicMock name='update_sources_list' id='139035065579536'>
required_packages_directive = [{'name': 'package'}]
expected_apt_update_calls = [call('/etc/apt/sources.list')]
expected_apt_install_calls = [call(['package'])], expected_result = True
concrete_entitlement_factory = <function concrete_entitlement_factory.<locals>.factory at 0x7e739fb45080>

    @pytest.mark.parametrize(
        [
            "required_packages_directive",
            "expected_apt_update_calls",
            "expected_apt_install_calls",
            "expected_result",
        ],
        [
            (
                None,
                [],
                [],
                True,
            ),
            (
                [],
                [],
                [],
                True,
            ),
            (
                [{"name": "package"}],
                [mock.call("/etc/apt/sources.list")],
                [mock.call(["package"])],
                True,
            ),
            (
                [{"name": "package"}, {"name": "package2"}],
                [mock.call("/etc/apt/sources.list")],
                [mock.call(["package", "package2"])],
                True,
            ),
        ],
    )
    @mock.patch("uaclient.apt.update_sources_list")
    @mock.patch("uaclient.apt.run_apt_install_command")
    @mock.patch("uaclient.apt.run_apt_update_command")
    @mock.patch(
        "uaclient.entitlements.base.UAEntitlement.entitlement_cfg",
        new_callable=mock.PropertyMock,
    )
    def test_handle_required_packages(
        self,
        m_entitlement_cfg,
        m_apt_update,
        m_apt_install,
        m_update_sources_list,
        required_packages_directive,
        expected_apt_update_calls,
        expected_apt_install_calls,
        expected_result,
        concrete_entitlement_factory,
    ):
        entitlement = concrete_entitlement_factory()
        m_entitlement_cfg.return_value = {
            "entitlement": {
                "directives": {"requiredPackages": required_packages_directive}
            }
        }
    
        assert expected_result == entitlement.handle_required_packages()
        assert [] == m_apt_update.call_args_list
>       assert (
            expected_apt_update_calls == m_update_sources_list.call_args_list
        )
E       AssertionError: assert [call('/etc/a...ources.list')] == [call('/etc/a...ntu.sources')]
E         At index 0 diff: call('/etc/apt/sources.list') != call('/etc/apt/sources.list.d/ubuntu.sources')
E         Use -v to get more diff

uaclient/entitlements/tests/test_base.py:1397: AssertionError
----------------------------- Captured stdout call -----------------------------
Updating standard Ubuntu package lists
Installing package
_ TestHandleRequiredPackages.test_handle_required_packages[required_packages_directive3-expected_apt_update_calls3-expected_apt_install_calls3-True] _

self = <uaclient.entitlements.tests.test_base.TestHandleRequiredPackages object at 0x7e73a201a610>
m_entitlement_cfg = <PropertyMock name='entitlement_cfg' id='139035066561552'>
m_apt_update = <MagicMock name='run_apt_update_command' id='139035065407120'>
m_apt_install = <MagicMock name='run_apt_install_command' id='139035065411472'>
m_update_sources_list = <MagicMock name='update_sources_list' id='139035066575056'>
required_packages_directive = [{'name': 'package'}, {'name': 'package2'}]
expected_apt_update_calls = [call('/etc/apt/sources.list')]
expected_apt_install_calls = [call(['package', 'package2'])]
expected_result = True
concrete_entitlement_factory = <function concrete_entitlement_factory.<locals>.factory at 0x7e739fb44180>

    @pytest.mark.parametrize(
        [
            "required_packages_directive",
            "expected_apt_update_calls",
            "expected_apt_install_calls",
            "expected_result",
        ],
        [
            (
                None,
                [],
                [],
                True,
            ),
            (
                [],
                [],
                [],
                True,
            ),
            (
                [{"name": "package"}],
                [mock.call("/etc/apt/sources.list")],
                [mock.call(["package"])],
                True,
            ),
            (
                [{"name": "package"}, {"name": "package2"}],
                [mock.call("/etc/apt/sources.list")],
                [mock.call(["package", "package2"])],
                True,
            ),
        ],
    )
    @mock.patch("uaclient.apt.update_sources_list")
    @mock.patch("uaclient.apt.run_apt_install_command")
    @mock.patch("uaclient.apt.run_apt_update_command")
    @mock.patch(
        "uaclient.entitlements.base.UAEntitlement.entitlement_cfg",
        new_callable=mock.PropertyMock,
    )
    def test_handle_required_packages(
        self,
        m_entitlement_cfg,
        m_apt_update,
        m_apt_install,
        m_update_sources_list,
        required_packages_directive,
        expected_apt_update_calls,
        expected_apt_install_calls,
        expected_result,
        concrete_entitlement_factory,
    ):
        entitlement = concrete_entitlement_factory()
        m_entitlement_cfg.return_value = {
            "entitlement": {
                "directives": {"requiredPackages": required_packages_directive}
            }
        }
    
        assert expected_result == entitlement.handle_required_packages()
        assert [] == m_apt_update.call_args_list
>       assert (
            expected_apt_update_calls == m_update_sources_list.call_args_list
        )
E       AssertionError: assert [call('/etc/a...ources.list')] == [call('/etc/a...ntu.sources')]
E         At index 0 diff: call('/etc/apt/sources.list') != call('/etc/apt/sources.list.d/ubuntu.sources')
E         Use -v to get more diff

uaclient/entitlements/tests/test_base.py:1397: AssertionError
----------------------------- Captured stdout call -----------------------------
Updating standard Ubuntu package lists
Installing package package2
=============================== warnings summary ===============================
uaclient/conftest.py:27
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:27: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture(scope="session", autouse=True)

uaclient/conftest.py:43
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:43: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture(scope="session", autouse=True)

uaclient/conftest.py:57
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:57: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture(scope="session", autouse=True)

uaclient/conftest.py:71
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:71: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture(scope="session", autouse=True)

uaclient/conftest.py:84
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:84: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture(scope="session", autouse=True)

uaclient/conftest.py:159
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:159: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture

uaclient/conftest.py:293
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/conftest.py:293: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture(autouse=True)

uaclient/api/u/pro/security/fix/_common/__init__.py:46
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/api/u/pro/security/fix/_common/__init__.py:46: DeprecationWarning: In 3.13 classes created inside an enum will not become a member.  Use the `member` decorator to keep the current behavior.
    class _Value:

uaclient/cli/tests/test_cli_api.py:11
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/cli/tests/test_cli_api.py:11: DeprecationWarning: invalid escape sequence '\['
    """\

uaclient/cli/tests/test_cli_collect_logs.py:18
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/cli/tests/test_cli_collect_logs.py:18: DeprecationWarning: invalid escape sequence '\['
    """\

uaclient/cli/tests/test_cli_security_status.py:18
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/cli/tests/test_cli_security_status.py:18: DeprecationWarning: invalid escape sequence '\['
    """\

uaclient/files/tests/test_data_types.py:32
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/files/tests/test_data_types.py:32: PytestCollectionWarning: cannot collect test class 'TestData' because it has a __init__ constructor (from: uaclient/files/tests/test_data_types.py)
    class TestData(DataObject):

uaclient/tests/test_gpg.py:10
  /home/paride/ubuntu/git/ubuntu-pro-client/uaclient/tests/test_gpg.py:10: PytestDeprecationWarning: @pytest.yield_fixture is deprecated.
  Use @pytest.fixture instead; they are the same.
    @pytest.yield_fixture

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED uaclient/entitlements/tests/test_base.py::TestHandleRequiredPackages::test_handle_required_packages[required_packages_directive2-expected_apt_update_calls2-expected_apt_install_calls2-True] - AssertionError: assert [call('/etc/a...ources.list')] == [call('/etc/a...nt...
FAILED uaclient/entitlements/tests/test_base.py::TestHandleRequiredPackages::test_handle_required_packages[required_packages_directive3-expected_apt_update_calls3-expected_apt_install_calls3-True] - AssertionError: assert [call('/etc/a...ources.list')] == [call('/etc/a...nt...
================= 2 failed, 2439 passed, 13 warnings in 29.73s =================
```

**Comment by orndorffgrant** [comment]
Thank you @paride! I believe I have addressed all of your comments

**Comment by paride** [comment]
This basically LGTM, but I'd like to have another look at the whole diff once https://github.com/canonical/ubuntu-pro-client/pull/2887 (which I approved) is merged and made part of this PR (which I assume is the plan, right?).

File: uaclient/livepatch.py [hunk 3]
```diff
@@ -131,6 +131,11 @@ def status() -> Optional[LivepatchStatusStatus]:
             [LIVEPATCH_CMD, "status", "--verbose", "--format", "json"]
         )
     except exceptions.ProcessExecutionError as e:
+        # only raise an error if there is a legitimate problem, not just lack
+        # of enablement
+        if re.match("Machine is not enabled.", e.stderr):
```
Comment (basak) [inline uaclient/livepatch.py:136]: Is `re.match` what you want? The '.' is "any character" and hasn't been escaped. Maybe `"Machine is not enabled" in e.stderr` better fits what you're actually trying to do here?

**Review by basak — COMMENTED** [review]
Reviewing wearing my SRU hat, according to https://wiki.ubuntu.com/UbuntuAdvantageToolsUpdates. I don't see anything that overly concerns me. A few talking points are below, but subject to discussion (no requirement to change unless further discussion concludes that we think we should).

Commit c6b06d40: I spotted that `ubuntu-advantage-tools.prerm` wasn't updated and thought I'd achieved something useful in this review. But then I updated the branch and saw that you've fixed it already. I guess not then :-)

Commit 1e13f591: Should this come with a migration path for existing files? Otherwise, do we have to wait for the timer to run before it has its permission changed? If so and the timer is disabled, is it possible that we're going to have some users who have an unexpected file permission going forwards? This wouldn't be robust because five releases from now we'll have completely forgotten that such a situation can arise. How confident are we that there will be no users left behind on old permissions after this update?

Commit 066a77a6: Using `lock.SpinLock` seemed alarming. But it turns out that this isn't a spinlock, just an ordinary retrying lock, since it does sleep in between retries. Consider fixing the name of the class please!

Commit 5b6d9abc: This relies on /etc/os-release to detect Noble. Is this guaranteed to have been updated before the ubuntu-advantage-tools.postinst runs? What if ubuntu-advantage-tools gets run first? Ah - this is happening via release-upgrades.d? Is this safe and preferred over using plain apt? For example, if a release upgrade breaks apt and the user fixes it up by hand, then do-release-upgrade never gets the opportunity to clean up after.

Risks

There's some churn in how apport and log collection work, and our interactions with apt (switch to deb822, migration path, etc). If there were a serious regression, these seem like the likely areas that could cause the most impact (eg. screwing up an upgrading user's apt configuration, or crashing when trying to send error reports). It might be worth considering what our test coverage is like in these areas (I'm not judging either way).

File: uaclient/livepatch.py (hunk 3 above)
Comment (orndorffgrant) [reply uaclient/livepatch.py:136]: I agree, we'll fix it before releasing.

**Comment by orndorffgrant** [comment]
Thank you @basak!!

> Reviewing wearing my SRU hat, according to https://wiki.ubuntu.com/UbuntuAdvantageToolsUpdates. I don't see anything that overly concerns me. A few talking points are below, but subject to discussion (no requirement to change unless further discussion concludes that we think we should).
> 
> Commit [c6b06d4](https://github.com/canonical/ubuntu-pro-client/commit/c6b06d4040ff37c78ec2a52b3b1bee3a8c12411c): I spotted that `ubuntu-advantage-tools.prerm` wasn't updated and thought I'd achieved something useful in this review. But then I updated the branch and saw that you've fixed it already. I guess not then :-)

:)

> Commit [1e13f59](https://github.com/canonical/ubuntu-pro-client/commit/1e13f59181599d437e89ec5ee18564a30a30a247): Should this come with a migration path for existing files? Otherwise, do we have to wait for the timer to run before it has its permission changed? If so and the timer is disabled, is it possible that we're going to have some users who have an unexpected file permission going forwards? This wouldn't be robust because five releases from now we'll have completely forgotten that such a situation can arise. How confident are we that there will be no users left behind on old permissions after this update?

If the timer is disabled or doesn't run and complete for some other reason, then the root-owned jobs-status.json file will have `600` permissions indefinitely. That means that effectively, #2601 would not be solved for those users; non-root `pro collect-logs` or `ubuntu-bug ubuntu-advantage-tools` would fail to collect `jobs-status.json`, and there would be some warnings in the log file about that case (Note that those warnings are not printed to stderr anymore). That is a small inconvenience while triaging bug reports, because in a scenario where we think the contents of jobs-status.json may matter, we'd have to ask the bug reporter to grab it manually.

The team, considering the (small) added risk of adding a postinst migration for this file, thought that this was an acceptable situation, and so decided not to add a postinst migration for the file permissions; however, if you think we might as well give it a postinst migration ("if upgrading from <31, then chmod 640 jobs-status.json"), we can do that.

> Commit [066a77a](https://github.com/canonical/ubuntu-pro-client/commit/066a77a6c97746a7b87ff5207cdff1631b5ac151): Using `lock.SpinLock` seemed alarming. But it turns out that this isn't a spinlock, just an ordinary retrying lock, since it does sleep in between retries. Consider fixing the name of the class please!

No time like the present - I'll fix this before releasing

> Commit [5b6d9ab](https://github.com/canonical/ubuntu-pro-client/commit/5b6d9abcb5b37c312d69194c6833c8e794890acc): This relies on /etc/os-release to detect Noble. Is this guaranteed to have been updated before the ubuntu-advantage-tools.postinst runs? What if ubuntu-advantage-tools gets run first? Ah - this is happening via release-upgrades.d? Is this safe and preferred over using plain apt? For example, if a release upgrade breaks apt and the user fixes it up by hand, then do-release-upgrade never gets the opportunity to clean up after.

Historically, some release-upgrade transitions are in postinst, but the more complex contract-related ones for attached users have gone in release-upgrades.d. I'm not sure which is really better.

I believe the worst case scenario if this migration doesn't run is that the user will get some messages from noble-APT after an attached upgrade from jammy/mantic => noble. But the .list will still work. The fix for a user would be to `pro disable` and `pro enable` the services affected.
@renanrodrigo what else could happen?

@basak do you have advice on where it is better to run this migration: via release-upgrades.d or via postinst?

> Risks
> 
> There's some churn in how apport and log collection work, and our interactions with apt (switch to deb822, migration path, etc). If there were a serious regression, these seem like the likely areas that could cause the most impact (eg. screwing up an upgrading user's apt configuration, or crashing when trying to send error reports). It might be worth considering what our test coverage is like in these areas (I'm not judging either way).

Ack, the changes to apport and log-collection are minor, and we have some integration tests in this area. We have some tests around the deb822 transition, and will at least do a release-upgrade test from jammy->noble during verification to check that the format is changed correctly. I'll add all of this to the SRU bug.

**Comment by orndorffgrant** [comment]
Hello @paride and @basak! This is now officially ready.

I've added the commits from the binary-package-rename branch that was reviewed separately. Those are commits [d39a017](https://github.com/canonical/ubuntu-pro-client/pull/2913/commits/d39a017a402bf9eb889704e315944b4b931f099a) through [5f44b02](https://github.com/canonical/ubuntu-pro-client/pull/2913/commits/5f44b029ff37456caf4b582ef3b9e4aa38773287). Feel free to re-review them to ensure they are the same as what you reviewed (I just cherry-picked them on to this branch).

There are a couple additional commits we needed to include in this release added at the end as well. In particular, [26ecd2c](https://github.com/canonical/ubuntu-pro-client/pull/2913/commits/26ecd2c57a9bf864d9e3f59a2f53f40b735ada2b) is worth taking a look at.

When you are both happy, then I believe @paride can go ahead and upload :)

**Comment by orndorffgrant** [comment]
I'm noticing some tests are red. I'll triage them before we upload to ensure they are just flakes, or fix them if needed.

**Comment by paride** [comment]
26ecd2c57a9bf864d9e3f59a2f53f40b735ada2b LGTM, and we're all green with the tests it seems, I think we're ready with this!

**Comment by orndorffgrant** [comment]
Great! Thank you @paride!

I've run a large set of our behave tests against a build in our staging ppa and found one bug where livepatch was failing to install when snapd was not up to date. I've added a fix in [ae56423](https://github.com/canonical/ubuntu-pro-client/pull/2913/commits/ae56423ae44140fc36f42347c80094441d2748f6). The other new commit is small updates to outdated tests and don't affect functionality.

Now we are ready for sponsor/upload :)

**Pull request merged**